### PR TITLE
USAGOV-908-snapshots-from-circleci: USAGOV-908 - Add site snapshot to cci

### PR DESCRIPTION
**Pull request by arpage**
## Jira Task

https://cm-jira.usa.gov/browse/USAGOV-908

## Description

Add snapshot backup approval and tasks for CircleCI.

Currently limited to this branch.

Also - the naming of the snapshots was envisioned to always include the name of the Jira ticket we are deploying from, simply for traceability (i.e. Is this a snapshot we care about? How do we grab the snapshots from the last deployment?).

Because we are always deploying code from the branch corresponding to the space to which we are deploying, there is no current way that I know of to add the Jira ticket ID into the snapshot name.

The snapshot tag is being currently constructed as:

`<< pipeline.git.branch >>.<< parameters.envname >>.<< pipeline.number >>` 

So when we are deploying from the USAGOV-908-snapshots-from-circleci branch, to dev the tag will be, for example:

`USAGOV-908-snapshots-from-circleci.dev.4850`

When we will be deploying to prod, the tag would be, for example:

`prod.prod.5726`

If there was a way to (git) tag the prod branch with the Jira ticket ID at the point we are deploying, we could possibly trigger just this workflow from the tag.

I'd really like to keep the snapshot tags linked to the deployment ticket IDs if at all possible.

That being said, this works now and the snapshot tag does include the CircleCI build ID - so that is something.

## Checklist for the Developer

- [x] A link to the JIRA ticket has been added above.
- [x] The JIRA ticket identifies the desired result of this change.
- [ ] The JIRA ticket contains clear acceptance criteria.
- [ ] The JIRA ticket contains clear testing steps.
- [ ] The JIRA ticket contains a description of "Done".
- [ ] Any preparation/installation/update steps required for this code to work properly (drush commands, scripts, configuration updates) are documented in the ticket.

## Checklist for the Peer Reviewers
- [ ] The branch name of this PR matches the project standards.
- [ ] QA steps are followed and any changes to process are updated in the ticket.
- [ ] Code Standards are followed, there are no bad practices in use.
- [ ] Config changes (if any) include only necessary changes.
- [ ] No errors in Drupal or Client Browser.
- [ ] Code has been tested locally (if possible).
- [ ] Following AC and Testing Steps verify changes work as expected.
- [ ] There are no known side-effects outside of expected behavior.

### CODE REVIEW

File: bin/snapshot-backups/site-snapshot-download
Comment (ghost) [inline bin/snapshot-backups/site-snapshot-download:1]: This was always downloading the current static site content from the current space.   Modified to actually download the specified backup tag.

File: bin/snapshot-backups/db-snapshot-download
Comment (ghost) [inline bin/snapshot-backups/db-snapshot-download:1]: Change what we are dowloading such that we don't try to append the space to the tag name we were given on the command line.

File: bin/snapshot-backups/db-dump-to-snapshot
Comment (ghost) [inline bin/snapshot-backups/db-dump-to-snapshot:1]: Wrap the db dump, dump validation, and pushing of dump to s3 into a single script.  Makes it easier to call from CircleCI also.

File: .circleci/config.yml
Comment (ghost) [inline .circleci/config.yml:1]: Add snapshot backup approval and tasks for CircleCI.

File: bin/snapshot-backups/site-snapshot-download [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env bash
 
 # we might be running in circleci (really?)
```
Comment (akf) [inline bin/snapshot-backups/site-snapshot-download]: You can take out the "(really?)" here -- that was me. Apparently we really might be.

**Review by akf — DISMISSED** [review]
This does seem to work! 
I haven't actually checked it out and tested downloading a snapshot using this version of db-snapshot-download. 

I have an idea about getting the deployment ticket ID into the tag. If we reliably enter the ticket ID as the first part of the pull request when doing a deployment merge, it should be available via "git show." Maybe we should hedge our bets by adding the commit hash as well -- that would give us something we could absolutely track back to a commit, even though it wouldn't be meaningful at a glance. 

For example (this is on an out-of-date checkout of prod): 

```
amykfarrell@FCOH2J-T6L7Q05P usagov-2021 % git show --pretty=%h    
addc7b148

amykfarrell@FCOH2J-T6L7Q05P usagov-2021 % git show --pretty=%b
USAGOV-874: post-cutover deployment
```

Hmm, might it be possible to get GitHub to require that we enter a well-formed Title on pull requests?

**Comment by ghost** [comment]
@akf and @jacobaaronyeager :

**Update to status of the Snapshot Tagging:**

I modified the CircleCI workflow to trigger the appropriate workflow (take snapshot of prod, stage or dev) when detecting formatted git tags,  This is completely separate from the branch workflow triggering, which is nice.

So, I'm envisioning the stage and prod deployment workflow to be modified to tag the appropriate branch, immediately after the deployment PR is merged, thereby tagging the code that will be deployed.  The git tag format will link it back to the Jira Deployment Ticket, and CircleCI will name the backup snapshot with a derivative that also links back to the deployment ticket.


The required git tag format can be seen on the included screenshot, and in `.circleci/config.yml` between lines 382 and 430.

The format of the snapshot tag generated by CircleCI can be seen at `.circleci/config.yml` lines 111 and 112.

USAGOV-908-CircleCI-Snapshot-Workflow

**Review by akf — APPROVED** [review]
I like it. I approved and ran the "snapshot dev" job to see it in action.

**Review by jacobaaronyeager — APPROVED** [review]

**Pull request merged**